### ISSUE TITLE
Support use third-party CA to replace sentry

### ISSUE THREAD

**Issue by daixiang0** [dapr/dapr#3968]
/area sentry

## Describe the proposal

From what I know, now `sentry` component works as a CA server and control mTLS between Dapr sidecars. 

Consider this case -- a Kubernetes cluster has API server, it works as a CA, setup service mesh like Istio, it also setup a CA, setup Dapr, another CA. Now we have 3 CA servers, maybe more due to business needs. Each CA has its own expire date but provides similar functions, maintaining them is very complex.

/cc @artursouza @yaron2 @wcs1only WDYT?

**Comment by yaron2** [comment]
Yes, I think having cert-manager support will be of great help, especially towards auto-rotating root certs.

**Comment by daixiang0** [comment]
Yes, now we need to rotate it by hand refer to https://docs.dapr.io/operations/security/mtls/#updating-root-or-issuer-certs

**Comment by ItalyPaleAle** [comment]
I really like this idea.

In the theme of supporting third-party CA's, I would like to suggest that we support Azure Key Vault as well, as it would allow better security (keys can't be extracted), automatic cert rotation, etc.

CC: @paulyuk

**Comment by ItalyPaleAle** [comment]
👋 bot

**Comment by daixiang0** [comment]
@safanaj I see you create a transformer as https://github.com/safanaj/dapr-cert-transformer, are you interested in introducing it into dapr, not a single project but an internal pkg?

**Comment by safanaj** [comment]
@daixiang0 I wrote that "transformer" as workaround because the filenames/keys for the TLS secret are constants, I would suggest to fix that instead to add a component, but also the latter approach could be ok, just I don't think I have time to work on a proper integration code now

**Comment by ItalyPaleAle** [comment]
@safanaj what does your code do? It wasn't clear from the README and I am not allowed to read your code because it's GPL'd 😔

**Comment by safanaj** [comment]
Hi @ItalyPaleAle , basically it just keep watching a secret (especially dapr-trust-bundle) for the keys "tls.key" and "tls.crt" and copy the content respectively to "issuer.key" and "issuer.crt" , this is pretty useful if you want the dapt-trust-bundle to be be managed by cert-manager, and provide it to dapr-sentry and all other components.

> I am not allowed to read your code because it's GPL'd 😔
what do you mean?

**Comment by ItalyPaleAle** [comment]
@safanaj thanks for explaining, makes sense now and I see its value - although I agree that we should probably be looking into an integrated solution rather than (yet another) background service.

As for the last part… because you released your source code under a GPL license, our policies require that I do not study/read your code, otherwise there's the risk that I may (even unconsciously) copy your code or parts of it. We would not want to risk getting the Dapr codebase tainted with GPL code, that would be really bad for a lot of people :)

**Comment by JoshVanL** [comment]
Hi all :wave:

I am a [cert-manager](https://github.com/cert-manager/cert-manager) maintainer, and would be keen to implement an integration :slightly_smiling_face: 

As I understand, this would involve adding a [`CertificateAuthority`](https://github.com/dapr/dapr/blob/bd96fea8d6ac77ad59013c5fbbbe1c1229471eec/pkg/sentry/ca/certificate_authority.go#L36) implementation for cert-manager?

`SignCSR` would then be a case of creating the corresponding [`CertificateRequest`](https://cert-manager.io/docs/concepts/certificaterequest/) resource for cert-manager to sign.

**Comment by yaron2** [comment]
> Hi all 👋
> 
> I am a [cert-manager](https://github.com/cert-manager/cert-manager) maintainer, and would be keen to implement an integration 🙂

That would be great and highly appreciated!

> As I understand, this would involve adding a [CertificateAuthority](https://github.com/dapr/dapr/blob/bd96fea8d6ac77ad59013c5fbbbe1c1229471eec/pkg/sentry/ca/certificate_authority.go#L36) implementation for cert-manager?

> SignCSR would then be a case of creating the corresponding [CertificateRequest](https://cert-manager.io/docs/concepts/certificaterequest/) resource for cert-manager to sign.

Your understanding is correct. If changes are needed to the interface or request/response objects, that's definitely doable so long as it doesn't involve breaking changes to existing Sentry CA users.

**Comment by JoshVanL** [comment]
Understood, thanks @yaron2. I’ll see what I can do over the next week or so to get more familiar with the project and give it a go. Cheers!

**Comment by ItalyPaleAle** [comment]
@JoshVanL this is great, thanks!

My only comment is that we probably don't want to enable support for Let's Encrypt in Dapr, since all endpoints are private. However, support for the ACME protocol (and things like Boulder) could certainly be welcome :)

(The next step .... how could we promote Azure Key Vault integration in cert-manager? :) )

**Comment by yaron2** [comment]
@ItalyPaleAle what's the concern with Let's Encrypt and public endpoints? we allow Dapr to fetch secrets from public endpoints today, so I'm not sure why fetching a cert would be any different.

**Comment by JoshVanL** [comment]
@ItalyPaleAle  Indeed publicly issued certs using ACME won’t work, however there is not a way to block that from the Dapr side easily. cert-manager uses untyped name references ([‘issuerRef’](https://cert-manager.io/docs/projects/istio-csr/#which-issuer-type)) to the issuer to sign the request. The best we can do is similar to [istio-csr](https://cert-manager.io/docs/projects/istio-csr/#which-issuer-type), which is a big warning that ACME issuers will not work.

> (The next step .... how could we promote Azure Key Vault integration in cert-manager? :) )

The best start would be to open up an issue on [GitHub](https://github.com/cert-manager/cert-manager) with exactly what you would like to see. I’m actually surprised to see it not have an issue already (AFAICS) since it has been mentioned on the kubernetes #cert-manager slack channel a few times.

**Comment by ItalyPaleAle** [comment]
@yaron2 

> what's the concern with Let's Encrypt and public endpoints? we allow Dapr to fetch secrets from public endpoints today, so I'm not sure why fetching a cert would be any different.

Two reasons:

1. It is unlikely to work, or at least not easy/convenient. The HTTP-01 and TLS-ALPN-01  challenges require each service to be directly addressable from the Internet. Although possible, I would not recommend users to expose every single service publicly. The DNS-01 challenge is the only one that could be usable, but it requires every service to have a DNS name that is part of a public domain. So it is a bit limited.
2. Second reason is privacy. Let's Encrypt maintains a record of all certificates that have been issues in the Certificate Transparency logs (https://letsencrypt.org/docs/ct-logs/), which are public and records cannot be removed from there. So everyone would be able to see for each domain what DNS entries exist (even if they're not routable, they can contain sensitive info).

In general LE is meant for public-facing services.

However the ACME protocol can be used with a private ACME server such as [boulder](https://github.com/letsencrypt/boulder) and that should work just fine :)

**Comment by ItalyPaleAle** [comment]
@JoshVanL as for implementation details. When I (briefly) looked into this the biggest challenge I found was in the data structure used to store the certs issued by Sentry.

Currently, Sentry expects exactly 3 files to be started:

https://github.com/dapr/dapr/blob/a4eefaed5e7f376f04968a1c6366a39a855ddff8/cmd/sentry/main.go#L93-L95

- "issuerCert" and "issuerKey" are the certificate and key for the intermediate CA that issues the certificates
- "rootCert" is the root CA cert (no key for this cert)

This is not very flexible, so for example you cannot pass a cert chain but only 1 root CA and 1 intermediate CA only. And both have to be present.

When Sentry is started with an external CA, then we store the CA cert as "rootCert" (note that it must be a root certificate), and then we request an intermediate cert and save that in issuerCert/issuerKey. The intermediate CA is the one used to generate leaf certs for services.

When Sentry is configured to generate self-signed certs, because it still has to produce those 3 files (and exactly 3), it:

1. Generates a root CA (self-signed), creating both the cert and key
2. It issues an intermediate CA signed by the root CA
3. It stores the intermediate CA as issuerCert/issuerKey. It stores the root CA cert as rootCert. It does **not** store the root CA key (and that's what causes issues such as the inability to renew the root CA cert, including consequences like this: https://github.com/dapr/dapr/issues/4237 )

I do think we need to be more flexible there, and in the case of ACME where there can be multiple intermediate certs, that's a requirement. So the biggest blocker here is refactor how Sentry generates certs and stores them.

I was thinking that we should store 2 things:

- An "issuerKey" that is the one that can be used to sign leaf certificates (and this can be a root key or an intermediate one)
- A "certChain" which contains the entire chain of certs

When using self-signed certs, "certChain" would be the root CA cert only, and "issuerKey" the root CA key only. If certs come from somewhere else, then things need to be adjusted accordingly.

This is the entire extent of my research so far. I have been meaning to look into this for a couple of months, but keep getting pulled into different directions. Perhaps you feel like addressing these issues too, and unblock adding third-party CAs? :)

**Comment by JoshVanL** [comment]
Thanks for the detailed explanation :smile: 

Indeed, this current way of doing this is incompatible with how cert-manager would do things- not least since we don't have direct access to the signing key.

I'll see how I get on, and keep in mind making sentry more generic to allow for other external CA implementations.

**Comment by irbekrm** [comment]
> (The next step .... how could we promote Azure Key Vault integration in cert-manager? :) )

Same as @JoshVanL would also be interested to see more what folks would want to gain from this integration.

We have looked at a few ways how AKV could work with cert-manager.
What we mostly get asked is whether there is some mechanism how to simply sync cert-manager issued certs from Kubernetes `Secret`s to AKV (for consumption by Azure services that don't consume certs from Kubernetes `Secret`s)

Another integration point could be using AKV just to generate and store private keys- we have in the past worked on a design for integration with external secret stores, would be keen to hear more use cases for this (i.e the advantages of this from security perspective).

There are also a few ways to integrate with AKV's cert creation mechanism- either via an external issuer (that would I guess mean DigiCert or GlobalSign certs as [those are CAs that AKV integrates with](https://docs.microsoft.com/en-us/azure/key-vault/certificates/create-certificate#partnered-ca-providers)) or via some mechanism that would integrate with [AKV's 'non-partnered' CAs mechanism](https://docs.microsoft.com/en-us/azure/key-vault/certificates/certificate-scenarios#creating-a-certificate-with-a-ca-not-partnered-with-key-vault) - so essentially getting the CSRs from AKV and getting signed by cert-manager's issuer.

**Comment by ItalyPaleAle** [comment]
@irbekrm thanks for your thoughts!

> Another integration point could be using AKV just to generate and store private keys- we have in the past worked on a design for integration with external secret stores, would be keen to hear more use cases for this (i.e the advantages of this from security perspective).

This was my first idea as well, and perhaps a "lower-hanging" (though possibly not "low-hanging") fruit. You could store your CA's keys into Azure Key Vault. The advantage of doing that is that signing operations would then be performed within the key vault: using ["sign" invocations](https://docs.microsoft.com/en-us/rest/api/keyvault/keys/sign), you can send AKV the plain-text (in this case the info about the certificate to sign) and  AKV responds with the digital signature. The signing operation is executed directly within the vault so the keys are never extracted, bringing a very interesting security benefit. Additionally, it's possible to enable auditing on AKV, so all signing operations would be recorded in an audit log too.

> There are also a few ways to integrate with AKV's cert creation mechanism- either via an external issuer (that would I guess mean DigiCert or GlobalSign certs as [those are CAs that AKV integrates with](https://docs.microsoft.com/en-us/azure/key-vault/certificates/create-certificate#partnered-ca-providers)) or via some mechanism that would integrate with [AKV's 'non-partnered' CAs mechanism](https://docs.microsoft.com/en-us/azure/key-vault/certificates/certificate-scenarios#creating-a-certificate-with-a-ca-not-partnered-with-key-vault) - so essentially getting the CSRs from AKV and getting signed by cert-manager's issuer.

And this is indeed the second thing I was thinking about, although perhaps more advanced.

**Comment by ItalyPaleAle** [comment]
👋 bot!

**Comment by daixiang0** [comment]
@JoshVanL are you still working on this issue? It is a great improvement to implement cert-manger.

**Comment by kiambogo** [comment]
bump on this – would be great to see this implemented :)

**Comment by lburgazzoli** [comment]
Within the context of https://github.com/dapr-sandbox/dapr-kubernetes-operator, having support for letting OpenShift injecting certificates using the [service serving certificate](https://docs.openshift.com/container-platform/4.13/security/certificates/service-serving-certificate.html) would be great

**Comment by jseguillon** [comment]
Hi there, this feature would be greatly appreciated. Could someone draw a path of actions to be taken to implement this ? Would love to help with that
